xla/service/gpu/gpu_hlo_cost_analysis.cc

```diff
@@ -316,12 +336,65 @@ int64_t FlopsPerElement(const absl::string_view device_name,
   return FindOrDefault(dtype_profiles->get(), opcode, kDefaultFlopsPerElement);
 }
 
+int64_t GetFlopsForElementwiseOp(const GpuDeviceInfo* gpu_device_info,
+                                 const HloOpcode op_code, const Shape& shape) {
+  int64_t flop_per_element = FlopsPerElement(
+      gpu_device_info ? gpu_device_info->name : kDefaultDeviceName,
+      shape.element_type(), op_code);
+  return flop_per_element * ShapeUtil::ElementsInRecursive(shape);
+}
+
+int64_t GetFlopsForElementwiseOp(const GpuDeviceInfo* gpu_device_info,
+                                 const HloInstruction* instr) {
+  LOG(ERROR) << "#####ShapeUtil::ElementsInRecursive(instr->shape()) "
+             << ShapeUtil::ElementsInRecursive(instr->shape())
+             << " INSTR: " << instr->ToString();
+
+  return GetFlopsForElementwiseOp(gpu_device_info, instr->opcode(),
+                                  instr->shape());
+}
+
+Status GpuHloCostAnalysis::HandleAllReduce(const HloInstruction* allreduce) {
+  const HloModuleConfig& config = allreduce->GetModule()->config();
+  int64_t num_devices = config.num_partitions();
+  int64_t num_replicas = config.replica_count();
+  // Data parallel ranks
+  int64_t num_ranks = num_devices;
+
+  int64_t output_bytes_accessed = 0;
+  ShapeUtil::ForEachSubshape(
+      allreduce->shape(), [&](const Shape& subshape, const ShapeIndex&) {
+        if (subshape.IsArray()) {
+          output_bytes_accessed += GetShapeSize(subshape);
+        }
+      });
+  int64_t bytes_accessed = output_bytes_accessed;
+  for (const HloInstruction* operand : allreduce->operands()) {
+    bytes_accessed += GetShapeSize(operand->shape());
+  }
+  current_properties_.set_output_bytes_accessed(output_bytes_accessed);
+  current_properties_[kBytesAccessedKey] = bytes_accessed;
+  current_properties_[kCollNumDevicesKey] = num_ranks;
+  // Since allreduce has compute, we need to get flops for the compute
+  // part which is an elementwise op.
+  current_properties_[kFlopsKey] = GetFlopsForElementwiseOp(
+      device_info_, allreduce->to_apply()->root_instruction()->opcode(),
+      allreduce->shape());
+
+  // Compute algorithmic scaling ratio, this can be used with link bandwidth
+  // to get the effective bandwidth of the algorithm.
+  // TODO TJ support multi-node case, we need to know how many nodes there are.
+  int num_intra_steps = 2 * (num_ranks - 1);
+  float scaling_ratio = (1.0 * num_ranks) / num_intra_steps;
```

Comment:
Some documentation here might be useful as to how the scaling ratio is transformed to actual ns timing.

Reply:
added some explanation